### PR TITLE
DataViews: add filters to table columns

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/55083
Follow-up to https://github.com/WordPress/gutenberg/pull/55270

## What?

This PR adds support for column filters:

https://github.com/WordPress/gutenberg/assets/583546/e05872a3-ba29-460f-b0ec-1ed382081162

## Testing Instructions

- Enable the "wp-admin" experiment and visit "Site editor > Pages > Manage all pages".
- Using the `list` type view, interact with the columns that have filters.

### CODE REVIEW

File: packages/edit-site/src/components/dataviews/view-list.js [hunk 1]
```diff
@@ -49,7 +52,8 @@ const sortingItemsInfo = {
 	desc: { icon: arrowDown, label: __( 'Sort descending' ) },
 };
 const sortIcons = { asc: chevronUp, desc: chevronDown };
-function HeaderMenu( { dataView, header } ) {
+// TODO: why doesn't Header work with view/onChangeView?
+function HeaderMenu( { dataView, header, view, onChangeView } ) {
```
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/view-list.js]: I've noticed the `HeaderMenu` doesn't take the `fields`, `view`, `onChangeView` as properties, but uses tanstack instead. Can we implement this component's functionality by only using our own abstractions?

Having to use two things for the same operations is confusing. For the demo, I chose to just pass `view` and `onChangeView`, but we need to decide how this works.

File: packages/edit-site/src/components/dataviews/view-list.js [hunk 2]
```diff
@@ -62,6 +66,41 @@ function HeaderMenu( { dataView, header } ) {
 	if ( ! isSortable && ! isHidable ) {
 		return text;
 	}
+	const isFilterable = header.column.columnDef.filters?.length > 0;
+	let filters = null;
+	if ( isFilterable ) {
+		filters = header.column.columnDef.filters.map( ( filter ) => {
```
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/view-list.js]: This and the UI components below is very similar to this code https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/dataviews/filters.js#L14

I wonder if the `Filters` component could render different children UI depending on the context (top-level vs columns, for example).

File: packages/edit-site/src/components/page-pages/index.js [hunk 3]
```diff
@@ -160,7 +160,19 @@ export default function PagePages() {
 						</a>
 					);
 				},
-				filters: [ 'enumeration' ],
+				filters: [
```
Comment (oandregal) [inline packages/edit-site/src/components/page-pages/index.js]: In terms of the filter API, I've added `columnLabel` as a filter property, though I don't think we should do this.

Note the labels for the "Author" filter depending on where is rendered (top-level or column filter) and whether it is a `IN` filter or a `NOT IN` one:

| Top level | Column filter |
| --- | --- |
| Captura de ecrã 2023-10-20, às 12 56 02 | Captura de ecrã 2023-10-20, às 12 57 34|

One way forward could be to make the filter type more granular. Instead of having only `enumeration`, we'll have `enumeration_in`, and `enumeration_not_in`. Using the same top-level label (`Author`) the column filters would add `is` or `is not` depending on the filter's type.

Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js]: Implemented `enumeration_in` and `enumeration_not_in` as filter types at https://github.com/WordPress/gutenberg/pull/55508/commits/f42b632646f4a34265f504227bbd9032dc0b282b

Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js]: In https://github.com/WordPress/gutenberg/pull/55508/commits/f42b632646f4a34265f504227bbd9032dc0b282b I've used the `enumeration_in` `enumeration_not_in` granurality to explore how to differentiate `IN` and `NOT IN` filters in the top-level. It could be as simple as having different UI components:

Captura de ecrã 2023-10-20, às 13 32 53

Though it's not yet implemented, the user may show/hide the filters they prefer, so no all filters need to be rendered at the same time.

Additionally, I'd recommend exploring other design options that show both `IN` and `NOT IN` filters using the same component. If we used a single component, at the API level, it'd be still valuable to have the filter type granularity provided by `enumeration_in`/`enumeration_not_in`. I don't see anything that would need to change API wise.

---

For testing this, apply this patch to your environment (add `author_exclude` to the list of `visibleFilters`and make sure you have the UI for the dataview CPT enabled so you can delete any dataview you have saved previously):

```diff
diff --git a/lib/experimental/data-views.php b/lib/experimental/data-views.php
index e46fe36ba0..7e9416ff24 100644
--- a/lib/experimental/data-views.php
+++ b/lib/experimental/data-views.php
@@ -15,7 +15,7 @@ function _gutenberg_register_data_views_post_type() {
                        'label'        => _x( 'Dataviews', 'post type general name', 'gutenberg' ),
                        'description'  => __( 'Post which stores the different data views configurations', 'gutenberg' ),
                        'public'       => false,
-                       'show_ui'      => false,
+                       'show_ui'      => true,
                        'show_in_rest' => true,
                        'rewrite'      => false,
                        'capabilities' => array(
diff --git a/packages/edit-site/src/components/dataviews/provider.js b/packages/edit-site/src/components/dataviews/provider.js
index bfde67a18a..2bbd922c3d 100644
--- a/packages/edit-site/src/components/dataviews/provider.js
+++ b/packages/edit-site/src/components/dataviews/provider.js
@@ -32,7 +32,7 @@ const DEFAULT_VIEWS = {
                        field: 'date',
                        direction: 'desc',
                },
-               visibleFilters: [ 'search', 'author', 'status' ],
+               visibleFilters: [ 'search', 'author', 'author_exclude', 'status' ],
                // All fields are visible by default, so it's
                // better to keep track of the hidden ones.
                hiddenFields: [ 'date', 'featured-image' ],
```

**Comment by oandregal** [comment]
@SaxonF @jameskoster What do you think we should do with filters that are not present at the top-level?

This is the use case: let's say a certain filter is not visible in the top level, but the column filter is. In this PR, you can see this at work by using the column filter "Author is not". If the user filters by "Author is not" which affordance should we use to communicate the filter status to the user?

I was thinking we could add a token-like element for this case, like in [this earlier exploration](https://github.com/WordPress/gutenberg/pull/55270#issuecomment-1761930866). It sounds it could work as it communicates the dataset is filtered, and the user can remove it from the top-level:

Captura de ecrã 2023-10-20, às 12 40 36

What would be your thoughts?

Note that this is also a question that we'd have to answer when the top-level filters can be hidden by the users. This is how to replicate in that scenario:

- "Filter A" is shown in the top-level bar.
- User filters the dataset by "Filter A".
- User hides the "Filter A" filter.
- At this point, the filter is hidden but the dataset is filtered by it.

**Comment by jameskoster** [comment]
Imo filters should always be visible if active, and hidden if inactive (except for primary filters).

We should also use the same UI, not a mix of dropdowns / chips.

The `is not` operator is a bit advanced, and we didn't finalise any designs for that yet. I'd suggest leaving it out until the designs have been explored as it's not clear where to place that control.

I put a prototype together that (hopefully) captures what we want to accomplish in this PR:


https://github.com/WordPress/gutenberg/assets/846565/fdaf34d6-b1cf-4360-ad11-5ef2e0095b1c

**Comment by jameskoster** [comment]
@oandregal Unsure if this is something to handle here or separately, but worth noting that some filters could have many terms, so we'll need scrollable lists, searching, and potentially loading states:


https://github.com/WordPress/gutenberg/assets/846565/6a1ec7d0-dba6-464b-81d0-50446648bc9b

**Comment by oandregal** [comment]
ok, I can limit this PR to just one filter per field and make it just about introducing column filters.

@jameskoster For the pages page, there are opportunities to have many filters per field: for example, Author ("in" and "not in") and Date ("before" and "after"). It'd be beneficial to have something for any of those soon to stress-test the filters API and make sure we're not missing anything.

**Comment by oandregal** [comment]
> Unsure if this is something to handle here or separately, but worth noting that some filters could have many terms, so we'll need scrollable lists, searching, and potentially loading states:

Absolutely. The design of the individual "Enumeration" filter component is something that needs a lot of work, and can be done in parallel.

**Comment by jameskoster** [comment]
Yes date is a particularly complex example. You might want to view pages published in a specific year, month, week, day, or time. Then you might want to view pages published before and/or after selected date(s) as you mentioned.

I'll work on some designs for that next. Fortunately Author and Status are a bit simpler so hopefully we can prioritise those.

**Comment by oandregal** [comment]
Interaction-wise, this has been limited to one filter of type `IN` for `author` and `status`. My next step is to make this code-ready.

https://github.com/WordPress/gutenberg/assets/583546/42cc995b-42aa-4366-b43e-2205295f3a5a

**Comment by jameskoster** [comment]
Nice! If possible as a part of this PR, it would be good to:

* Keep the menu open when selecting filters so you can multi-select quickly.
* Append a "Reset" button once selections have been made.

**Comment by oandregal** [comment]
> Nice! If possible as a part of this PR, it would be good to:

I'd like to address those as follow-ups. They can be worked on parallel and be picked up by anyone. UI is going to need a lot of focused efforts, and I'd rather have the basic interactions working first.

**Comment by oandregal** [comment]
This is mostly ready, both code & interaction wise. I'm fixing a subtle bug with the author column filter: while it works, the "active state" (the check for the selected item) is not set when selecting an author from the top-level filter. I've tracked it down to the top-level author component, that is converting every element to string instead or retaining the original type (integer).

File: packages/edit-site/src/components/dataviews/view-list.js (hunk 1 above)
Comment (oandregal) [reply packages/edit-site/src/components/dataviews/view-list.js]: Set to use the existing `dataView`, `header`. If this needs change, it'd be done in a different PR.

**Comment by oandregal** [comment]
@jameskoster @SaxonF The existing `filter` [icon](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library) is for CSS filters (duotone). Is there a SVG for data filtering somewhere? It can be added in a follow-up if it takes time.

**Comment by jameskoster** [comment]
@oandregal if possible could we add this icon to the library, potentially named `funnel` since `filter` is taken?

```
<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
<path d="M10 17.5H14V16H10V17.5ZM6 6V7.5H18V6H6ZM8 12.5H16V11H8V12.5Z" />
</svg>
```

**Comment by oandregal** [comment]
> if possible could we add this icon to the library, potentially named funnel since filter is taken?

Thanks! It was added as `funnel` at https://github.com/WordPress/gutenberg/pull/55508/commits/7703a3da0043f10e104c16b5c161eb1c9af192f4.

Captura de ecrã 2023-11-02, às 17 27 11

**Comment by oandregal** [comment]
I found an issue that's solved by https://github.com/WordPress/gutenberg/pull/55839

**Comment by oandregal** [comment]
Rebased to solve some issues. Would welcome a review @youknowriad @jorgefilipecosta @ntsekouras :)

File: packages/edit-site/src/components/dataviews/view-list.js [hunk 4]
```diff
@@ -119,6 +146,92 @@ function HeaderMenu( { dataView, header } ) {
...
+												);
+											} else {
+												dataView.setColumnFilters( [
+													...otherFilters,
+													{
+														[ filter.id + ':in' ]:
+															element.value,
+													},
+												] );
+											}
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/view-list.js:225]: Something in the formatting is wrong here right? Make it hard to follow

File: packages/edit-site/src/components/dataviews/view-list.js [hunk 5]
```diff
@@ -63,6 +67,29 @@ function HeaderMenu( { dataView, header } ) {
 		return text;
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/view-list.js:67]: Just a small non blocking remark: This component `HeaderMenu` is built as a tan stack component (gets dataView and header) and renders a dropdown. I wonder if it could be similar to built it as a DataView component instead. (get the view and field as props).

**Review by youknowriad — COMMENTED** [review]
This is working great overall but I noticed that we should reset the "page" once filters change.

**Comment by jameskoster** [comment]
> reset

That reminds me – we're missing a global reset button for filters.

Screenshot 2023-11-06 at 14 46 56

Appreciate it's one to handle separately but wanted to leave a note anyway.

**Comment by oandregal** [comment]
> This is working great overall but I noticed that we should reset the "page" once filters change.

Nice one, fixed at https://github.com/WordPress/gutenberg/pull/55508/commits/a9c651bf838054e8596498d9198058b71caa03c3

File: packages/edit-site/src/components/dataviews/view-list.js [hunk 6]
```diff
@@ -427,12 +427,13 @@ function ViewList( {
 			} );
 		},
 		onGlobalFilterChange: ( value ) => {
-			onChangeView( { ...view, search: value, page: 0 } );
+			onChangeView( { ...view, search: value, page: 1 } );
```
Comment (oandregal) [inline packages/edit-site/src/components/dataviews/view-list.js:430]: Apparently, this value is not being used – or is ignored once the own `Search` component updates it to 1 [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/dataviews/search.js#L25). I thought I'd change it anyway, though this needs a separate further investigation.

File: packages/edit-site/src/components/dataviews/view-list.js (hunk 5 above)
Comment (oandregal) [reply packages/edit-site/src/components/dataviews/view-list.js:67]: I had the same thought https://github.com/WordPress/gutenberg/pull/55508#discussion_r1366811917

**Comment by oandregal** [comment]
> Appreciate it's one to handle separately but wanted to leave a note anyway.

Would you think we could use the existing https://github.com/WordPress/gutenberg/issues/55100 for tracking filter-related work? I can help to maintaining it up to date as we go, if you want.

**Review by youknowriad — APPROVED** [review]
LGTM 👍 

Keeping track of all the remaining filters work on the dedicated issue feels like a great approach to me.

**Comment by oandregal** [comment]
I've updated the existing filter issue to track ongoing work https://github.com/WordPress/gutenberg/issues/55100 Hopefully, each task comment is self-explanatory (which ones need design, which need component work, etc.).
